### DISCUSSION TITLE
Localization not working in Browser using .NET 8

### DISCUSSION THREAD

**Discussion by andrewkdci** [AvaloniaUI/Avalonia#15223] [Q&A] [answered]
### Describe the bug

Localization following the guide [Localizing using ResX](https://docs.avaloniaui.net/docs/guides/implementation-guides/localizing) does not work in the browser when targeting .NET 8.

### To Reproduce

Reference the [Localizing using ResX](https://docs.avaloniaui.net/docs/guides/implementation-guides/localizing) guide to create a localization project.

1) Create an empty Avalonia project with a Desktop and Browser variant.
2) Update the Avalonia library, Avalonia.Desktop and Avalonia.Browser projects to target .NET 8.
3) Add `Lang\Resources.resx` and `Lang\Resources.ja-JP.resx` files to the Avalonia library.
4) In the Avalonia library .csproj file, update the `EmbeddedResource` entry for the `Resources.resx` file to use the `PublicResXFileCodeGenerator`:
```
  <ItemGroup>
    <EmbeddedResource Update="Lang\Resources.resx">
      <Generator>PublicResXFileCodeGenerator</Generator>
      <LastGenOutput>Resources.Designer.cs</LastGenOutput>
    </EmbeddedResource>
  </ItemGroup>
```

5) Regenerate `Resources.resx` using the "Run Custom Tool" option from the the context menu in Visual Studio's Solution Explorer. The generated `Resources.Designer.cs` should now have a `public` entry for the `Test` string, e.g.
```
        public static string Test {
            get {
                return ResourceManager.GetString("Test", resourceCulture);
            }
        }
```

6) Create a "Test" string in each resources file. In the default, make the value "english". In the ja-JP, make the value "japanese".
7) Update `App.xaml.cs` to the following:
```
using Avalonia;
using Avalonia.Controls.ApplicationLifetimes;
using Avalonia.Data.Core.Plugins;
using Avalonia.Markup.Xaml;
using AvaloniaLocalization.ViewModels;
using AvaloniaLocalization.Views;

namespace AvaloniaLocalization;

public partial class App : Application
{
    public override void Initialize()
    {
        AvaloniaXamlLoader.Load(this);
    }

    public override void OnFrameworkInitializationCompleted()
    {
        Lang.Resources.Culture = new System.Globalization.CultureInfo("ja-JP");

        // Line below is needed to remove Avalonia data validation.
        // Without this line you will get duplicate validations from both Avalonia and CT
        BindingPlugins.DataValidators.RemoveAt(0);

        if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
        {
            desktop.MainWindow = new MainWindow
            {
                DataContext = new MainViewModel()
            };
        }
        else if (ApplicationLifetime is ISingleViewApplicationLifetime singleViewPlatform)
        {
            singleViewPlatform.MainView = new MainView
            {
                DataContext = new MainViewModel()
            };
        }

        base.OnFrameworkInitializationCompleted();
    }
}
```

8) Update `MainView.xaml` to the following:
```
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:vm="clr-namespace:AvaloniaLocalization.ViewModels"
             xmlns:lang="clr-namespace:AvaloniaLocalization.Lang"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="AvaloniaLocalization.Views.MainView"
             x:DataType="vm:MainViewModel">
  <Design.DataContext>
    <!-- This only sets the DataContext for the previewer in an IDE,
         to set the actual DataContext for runtime, set the DataContext property in code (look at App.axaml.cs) -->
    <vm:MainViewModel />
  </Design.DataContext>

  <TextBlock Text="{x:Static lang:Resources.Test}" HorizontalAlignment="Center" VerticalAlignment="Center"/>
</UserControl>
```

### Expected behavior

For both desktop and browser variants, the app should display the word "japanese" in the middle of the screen.

This works for Desktop.

This does not work for Browser. In the browser, it shows the default value "english".

### Avalonia version

11.0.10

### OS

Windows 10 v19045.3693, x64
WebAssembly
Google Chrome v123.0.6312.105, 64-bit

### Additional context

Localization works for .NET 7 on the browser.

**Comment by maxkatz6** [comment]
Avalonia doesn't manage System.Globalization on any platform. You might need to check documentation on that from Microsoft.
Only thing that can be related to Avalonia, older versions might have forced invariant globalization, which can be fixed by using newer 11.1.0-beta1.

**Reply by andrewkdci** [reply]
Please explain to me what this article is saying, then: [Localizing using ResX](https://docs.avaloniaui.net/docs/guides/implementation-guides/localizing).

Sounds to me like localization is supported by Avalonia, is it not? What am I missing here?

**Reply by andrewkdci** [reply]
Furthermore, I don't understand why #15220 was closed out in favor of this discussion. 

Localization is supported in Avalonia. Localization works on Desktop and Browser for .NET 7. Localization does work on Desktop for .NET 8, but not on Browser for .NET 8. Please explain to me why this is not a bug.

**Reply by andrewkdci** [reply]
Ah, ok, I think I see what you're saying. Localization is not actually an Avalonia functionality. It's all via .NET. I can see that now.

However, Avalonia does provide the framework to convert a desktop into a WASM app, does it not? Therefore is it not within Avalonia's responsibility to convert that localization functionality to WASM too for .NET 8, especially since it does it successfully for .NET 7?

**Reply by andrewkdci** [reply]
Ok, I can see where this is going. I'm going to investigate whether localization is broken on .NET 8 (non-Avalonia) WASM apps. No need to pursue this issue further until I confirm this.

Thanks.

**Reply by andrewkdci** [reply]
Ok, I created two vanilla Blazor WASM apps, one targeting .NET 7 and one targeting .NET 8. I created the same Resources.resx files recommended by the Avalonia article. Per Microsoft's recommendation, in `Program.Main()` I set the culture statically before the `RunAsync` call, e.g.

```
            CultureInfo.DefaultThreadCurrentCulture = new CultureInfo("ja-JP");
            CultureInfo.DefaultThreadCurrentUICulture = new CultureInfo("ja-JP");

            await builder.Build().RunAsync();
```

In the markup, I used `@BlazorLocalizationNet7.Lang.Resources.Test` to print out the localized string.

It worked fine for both .NET 7 and .NET 8.

I can therefore only conclude that something in Avalonia is breaking WASM localization for .NET 8. Or am I still missing something?

**Reply by MrJul** [reply, marked as answer]
By default, the .NET WASM runtime doesn't load any satellite (resource) assemblies.
Blazor loads them for the current culture at startup (and doesn't support dynamic culture by default).

I don't believe it's Avalonia role to load them. You'll get the same behavior with a non-Blazor WASM project.

Workaround:

```csharp
internal partial class WebAssemblyCultureProviderInterop
{
    [JSImport("INTERNAL.loadSatelliteAssemblies")]
    public static partial Task LoadSatelliteAssemblies(string[] culturesToLoad);
}
```

Before using the culture:
```csharp
await WebAssemblyCultureProviderInterop.LoadSatelliteAssemblies(["ja-JP"]);
```

Only tested in .NET 8.0 (it might be different in .NET 7.0).
Source reference:

https://github.com/dotnet/aspnetcore/blob/16ae0ab9c61c994048513cd8fa3127608b7ec936/src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyCultureProvider.cs#L99-L103

**Reply by andrewkdci** [reply]
Thank you. Can you please expand on how to use this workaround in an Avalonia project? I can't seem to find `WebAssemblyCultureProviderInterop` anywhere. As far as I can tell, my Avalonia Browser head doesn't have access to any ASP.NET frameworks. Is there a NuGet package that I need?

**Reply by andrewkdci** [reply]
Nevermind. I got it!

In my Avalonia.Browser project, I modified Program.cs:
```
using System.Runtime.InteropServices.JavaScript;
using System.Runtime.Versioning;
using System.Threading.Tasks;

using Avalonia;
using Avalonia.Browser;

using AvaloniaLocalization;

[assembly: SupportedOSPlatform("browser")]

internal partial class Program
{
    private static async Task Main(string[] args)
    {
        await WebAssemblyCultureProviderInterop.LoadSatelliteAssemblies(["ja-JP"]);

        await BuildAvaloniaApp()
            .WithInterFont()
            .StartBrowserAppAsync("out");
    }

    public static AppBuilder BuildAvaloniaApp()
        => AppBuilder.Configure<App>();
}

internal partial class WebAssemblyCultureProviderInterop
{
    [JSImport("INTERNAL.loadSatelliteAssemblies")]
    public static partial Task LoadSatelliteAssemblies(string[] culturesToLoad);
}
```

I now see the expected localization.

Perhaps there's a more elegant way to implement this workaround. But for now this resolves the issue of .NET 8 breaking my WASM localization. Thanks!

**Reply by xuxiaokang0723** [reply]
When I try to publish in browser mode, it keeps giving an error message: Can't find the original satellite assembly in the list of resolved files to publish for asset。Do you know what's going on?
https://private-user-images.githubusercontent.com/178752908/398344655-462f32e0-8875-4dd2-b5a0-c027b2bed7cd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxMDg2ODksIm5iZiI6MTczNTEwODM4OSwicGF0aCI6Ii8xNzg3NTI5MDgvMzk4MzQ0NjU1LTQ2MmYzMmUwLTg4NzUtNGRkMi1iNWEwLWMwMjdiMmJlZDdjZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyNVQwNjMzMDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZjUzMDNmMWY4M2RiZWNlZjJhYWJhNDJmOGUwMGEzMzI4ZWNmYWE2Mzc2NjZhMWVhMjQ1YmI4YmE1NzcxYzBmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.-EVneaesAeqDFv7C00IRjVQwUqZCeOVgCneXWX8KfOM
https://private-user-images.githubusercontent.com/178752908/398344889-d406b1e8-bf59-4d4e-bd9b-f06abe8bae59.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxMDg2ODksIm5iZiI6MTczNTEwODM4OSwicGF0aCI6Ii8xNzg3NTI5MDgvMzk4MzQ0ODg5LWQ0MDZiMWU4LWJmNTktNGQ0ZS1iZDliLWYwNmFiZThiYWU1OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyNVQwNjMzMDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMDFlM2E3MjVmMjU2MDI3M2Y2YjU3YTE1YjllODA4NTQzNDY3YmJkYWI3NDg0YTgzZjEyODU4YzNkY2Q2OWRlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.f2FHjYW5hcXcBhZq3t5kyzIOseiqpCQS7yLOxtQnThE
